pageserver/src/pgdatadir_mapping.rs

```diff
@@ -312,46 +310,22 @@ impl Timeline {
 
             let key = rel_block_to_key(*tag, *blknum);
 
-            if ctx.has_perf_span() {
-                perf_instrument = true;
-            }
-
             let key_slots = keys_slots.entry(key).or_default();
             key_slots.push((response_slot_idx, ctx));
-        }
 
-        let keyspace = {
-            // add_key requires monotonicity
-            let mut acc = KeySpaceAccum::new();
-            for key in keys_slots
-                .keys()
-                // in fact it requires strong monotonicity
-                .dedup()
-            {
-                acc.add_key(*key);
-            }
-            acc.to_keyspace()
-        };
+            let acc = req_keyspaces.entry(lsn).or_default();
+            acc.add_key(key);
+        }
```

Comment:
What about the monotonicity requirement of add_key? (See the code you deleted to the left if you're using split diff view)